### PR TITLE
improved rollback logic

### PR DESCRIPTION
**Pull request by pwinckles**
**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3350

# What does this Pull Request do?

1. Exceptions during rollbacks are logged; not rethrown
2. Staging directories are always cleaned up
3. General logic and logging cleanup 

# How should this be tested?

Start F6:

```
mvn -Dfcrepo.home=target/fcrepo -Dfcrepo.log=DEBUG clean jetty:run
```

Create a resource in a transaction:

```
curl -u fedoraAdmin:fedoraAdmin -v -XPOST http://localhost:8080/rest/fcr:tx
txid=<TX_ID>
curl -u fedoraAdmin:fedoraAdmin -H"Atomic-ID: ${txid}" http://localhost:8080/rest/foo -H "Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel=type" -v -H "Content-Type: text/plain" -X PUT --data-binary "bar"
```

See the staging directory:

```
ls target/fcrepo/data/staging
```

Commit the transaction:

```
curl -u fedoraAdmin:fedoraAdmin -X PUT -v http://localhost:8080/rest/fcr:tx/${txid}
```

Verify the staging directory is now gone.

Do the same steps a second time, but, instead of committing the transaction, wait for it to expire and the confirm that the staging directory was removed.

# Interested parties

@fcrepo4/committers

### CODE REVIEW

**Comment by awoods** [comment]
As a follow-on ticket, we will likely want to catch `PersistentStorageException`s and send an appropriate response code to the client... instead of a stacktrace.

The following script demonstrates the issue:
```
#!/usr/bin/bash

x=0; 
r="http://localhost:8080/rest/"; 
while true; 
  do x=$(($x + 1));
  r=`curl -s -H"Slug: $x" -XPOST $r`; 
  echo "r = $r"; 
done 
```

**Comment by awoods** [comment]
Likely a non-related issue, but the following test results in a `org.springframework.dao.DuplicateKeyException`.

transactions.sh:
```
#!/usr/bin/bash

if [ "$1" == "" ]; then
    echo "There must be a txn-URI argument!"
    exit 1;
fi

tx="$1"
x=0;
r="localhost:8080/rest"
while true; 
  do x=$(($x + 1));
  curl -XPUT -H"Atomic-ID: $tx" "$r/$x"  -i
done 
```

1. Start a transaction:
   ```
   curl -i -XPOST http://localhost:8080/rest/fcr:tx
   ```
1. Pass the transaction id (`Location` header) into the `transactions.sh` script
   - let the script run for a few hundred iterations
1. Start a new transaction and a new run of `transactions.sh`, before the first transaction expires
   - BOOM!

**Comment by awoods** [comment]
Staging clean-up and rollback appear to work as expected.

**Comment by pwinckles** [comment]
> Likely a non-related issue, but the following test results in a `org.springframework.dao.DuplicateKeyException`.
> 
> transactions.sh:
> 
> ```
> #!/usr/bin/bash
> 
> if [ "$1" == "" ]; then
>     echo "There must be a txn-URI argument!"
>     exit 1;
> fi
> 
> tx="$1"
> x=0;
> r="localhost:8080/rest"
> while true; 
>   do x=$(($x + 1));
>   curl -XPUT -H"Atomic-ID: $tx" "$r/$x"  -i
> done 
> ```
> 
>     1. Start a transaction:
>        ```
>        curl -i -XPOST http://localhost:8080/rest/fcr:tx
>        ```
> 
>     2. Pass the transaction id (`Location` header) into the `transactions.sh` script
>        
>        * let the script run for a few hundred iterations
> 
>     3. Start a new transaction and a new run of `transactions.sh`, before the first transaction expires
>        
>        * BOOM!

This is will be fixed in PR #1710

**Comment by pwinckles** [comment]
> As a follow-on ticket, we will likely want to catch `PersistentStorageException`s and send an appropriate response code to the client... instead of a stacktrace.
> 
> The following script demonstrates the issue:
> 
> ```
> #!/usr/bin/bash
> 
> x=0; 
> r="http://localhost:8080/rest/"; 
> while true; 
>   do x=$(($x + 1));
>   r=`curl -s -H"Slug: $x" -XPOST $r`; 
>   echo "r = $r"; 
> done 
> ```
> 
> ..noting that getting the root cause based on this PR is an improvement over the previous functionality of getting a masked exception.

Yes, there is an existing Jira ticket for dealing with those exceptions

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/IndexBuilderImpl.java [hunk 1]
```diff
@@ -190,4 +210,32 @@ private boolean repoContainsObjects() {
...
+        try {
+            return Files.createDirectories(sessionStagingRoot.resolve(txId));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void cleanupStaging(final Path stagingDir) {
+        if (!FileUtils.deleteQuietly(stagingDir.toFile())) {
+            LOGGER.warn("Failed to cleanup staging directory: " + stagingDir);
```
[the remark this answers is not in the document]
Comment (pwinckles) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/IndexBuilderImpl.java:223]: Yup. I'll fix those

File: fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/ContainmentIndex.java [hunk 2]
```diff
@@ -39,7 +38,7 @@
      * @param fedoraResource The containing fedora resource
      * @return A stream of contained identifiers
      */
-    Stream<String> getContains(Transaction tx, FedoraResource fedoraResource);
+    Stream<String> getContains(String tx, FedoraResource fedoraResource);
```
[the remark this answers is not in the document]
Comment (pwinckles) [reply fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/ContainmentIndex.java:41]: Will do. Renamed the implementation, but forgot the interface.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/IndexBuilderImpl.java [hunk 3]
```diff
@@ -89,80 +97,92 @@ public void rebuild() {
         containmentIndex.reset();
         searchIndex.reset();
 
+        final var txId = UUID.randomUUID().toString();
```
[the remark this answers is not in the document]
Comment (pwinckles) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/IndexBuilderImpl.java:100]: Hmm... I'm not sure I understand. The whole point of the this class is the `rebuild` method, and there are tests that use it. `rebuildIfNecessary` is just a wrapper that tries to make a guess as to whether or not `rebuild` should be called.

[the remark this answers is not in the document]
Comment (awoods) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/IndexBuilderImpl.java:100]: `rebuild()` is never called outside of this class.

[the remark this answers is not in the document]
Comment (pwinckles) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/IndexBuilderImpl.java:100]: Ah, you're right. I used to be called by `RebuildIT`, but I changed that earlier this week. Sure, I can remove it from the interface. Presumably we might want it add it back later if we add the ability to force a rebuild?

**Comment by awoods** [comment]
> `PersistentStorageException`s

Do you happen to have a link to an existing JIRA ticket for this? I am not seeing it.

**Comment by pwinckles** [comment]
> `PersistentStorageException`s

You're right again. I was thinking of this, different, exception that we'd discussed earlier in the week.

https://jira.lyrasis.org/browse/FCREPO-3261

Seems like there should just be a ticket to review the exception handling to make sure it's generally not letting anything through.

**Pull request merged**